File: 24.BinarySearchVariants.py
# Lower Bound: The index of the first element that is ≥ target in a sorted list.
# Upper Bound: The index of the first element that is > target in a sorted list.

# Explanation

# Binary Search can be tweaked to find not only whether an element exists, 
# but also to find the position where it could be inserted while maintaining order:

# Lower Bound helps find the leftmost suitable position.

# Upper Bound helps find the rightmost suitable position (just after last occurrence).

# This is extremely useful in problems like:

# Counting occurrences of an element.

# Finding insertion points.

# Range queries (e.g., count of numbers ≤ X or ≥ X).

# We will write two functions using binary search logic.

def lower_bound(arr, target):
    low, high = 0, len(arr)
    while low < high:
        mid = (low + high) // 2
        if arr[mid] < target:
            low = mid + 1
        else: 
            high = mid
    return low

def upper_bound(arr, target):
    low, high = 0, len(arr)
    while low < high:
        mid = (low + high) // 2
        if arr[mid] <= target:
            low = mid + 1
        else:
            high = mid
    return low

arr = [1, 2, 4, 4, 4, 6, 7, 9]
target = 4

print("Array:", arr)
print("Target:", target)

lb = lower_bound(arr, target)
ub = upper_bound(arr, target)

print("Lower Bound index:", lb)
print("Upper Bound index:", ub)

File: 24.BinarySearchVariants.java
/* ----------------------------------------------------------------------------  */
/*   ( The Authentic JS/JAVA CodeBuff )
 ___ _                      _              _ 
 | _ ) |_  __ _ _ _ __ _ __| |_ __ ____ _ (_)
 | _ \ ' \/ _` | '_/ _` / _` \ V  V / _` || |
 |___/_||_\__,_|_| \__,_\__,_|\_/\_/\__,_|/ |
                                        |__/ 
 */
/* --------------------------------------------------------------------------   */
/*    Youtube: https://youtube.com/@code-with-Bharadwaj                        */
/*    Github : https://github.com/Manu577228                                  */
/*    Portfolio : https://manu-bharadwaj-portfolio.vercel.app/portfolio      */
/* -----------------------------------------------------------------------  */

import java.io.*;
import java.util.*;

public class BinarySearchVariants {

    // lower_bound: first index where arr[i] >= target
    static int lowerBound(int[] arr, int target) {
        int low = 0, high = arr.length;
        while (low < high) {
            int mid = (low + high) / 2;
            if (arr[mid] < target) {
                // if mid value < target, discard left part
                low = mid + 1;
            } else {
                // if arr[mid] >= target, possible answer, move left
                high = mid;
            }
        }
        return low;
    }

    // upper_bound: first index where arr[i] > target
    static int upperBound(int[] arr, int target) {
        int low = 0, high = arr.length;
        while (low < high) {
            int mid = (low + high) / 2;
            if (arr[mid] <= target) {
                // if mid value <= target, go right
                low = mid + 1;
            } else {
                // if arr[mid] > target, possible answer, move left
                high = mid;
            }
        }
        return low;
    }

    public static void main(String[] args) throws Exception {
        int[] arr = {1, 2, 4, 4, 4, 6, 7, 9};
        int target = 4;

        System.out.println("Array: " + Arrays.toString(arr));
        System.out.println("Target: " + target);

        int lb = lowerBound(arr, target);
        int ub = upperBound(arr, target);

        System.out.println("Lower Bound index: " + lb);
        System.out.println("Upper Bound index: " + ub);
    }
}
